src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets

```diff
@@ -136,6 +136,9 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Include="-static-pie" Condition="'$(StaticExecutable)' == 'true' and '$(PositionIndependentExecutable)' != 'false'" />
       <LinkerArg Include="-dynamiclib" Condition="'$(_IsApplePlatform)' == 'true' and '$(NativeLib)' == 'Shared'" />
       <LinkerArg Include="-shared" Condition="'$(_IsApplePlatform)' != 'true' and '$(NativeLib)' == 'Shared'" />
+      <!-- source-build links directly with openssl (-lssl -lcrypto) -->
```

Comment:
I assume that you will want to have libicu in this list as well. Is that correct?

I do not think that we want to pass in these libraries to the linker by default. We will need to have a new property for non-portable build output and set it as necessary.

What do you want the default for `dotnet publish` to be in distro-provided `dotnet`? Do you want it to produce the non-portable output default or portable output by default? This is general question for all targets, not specific to native AOT. Does it produce portable or non-portable output today?

Reply:
In the first PropertyGroup, we can explicitly define an internal property:

```xml
<_IsSourceBuild Condition="'$(NETCoreSdkRuntimeIdentifier)' != '$(NETCoreSdkPortableRuntimeIdentifier)'">true</_IsSourceBuild>
```

After that, make decisions based on that property: `Condition="'$(_IsSourceBuild)' == 'true'"`; without locking  down to linux.

Reply:
In the first PropertyGroup, we can explicitly define an internal property:

```xml
<_NonPortableBuild Condition="'$(NETCoreSdkRuntimeIdentifier)' != '$(NETCoreSdkPortableRuntimeIdentifier)'">true</_NonPortableBuild>
```

After that, make decisions based on that property: `Condition="'$(_NonPortableBuild)' == 'true'"`; without locking  down to linux.

Reply:
> What do you want the default for dotnet publish to be in distro-provided dotnet? Do you want it to produce the non-portable output default or portable output by default? This is general question for all targets, not specific to native AOT. Does it produce portable or non-portable output today?

Since non-portable and portable have different traits, we want this to be a deliberate choice by the user.

Similar to choosing between the .NET bundled runtime packs or the portable Microsoft NuGet packages, the UX for this is the user specify the non-portable rid (instead of the portable rid).

We should then add the linker flags based on the selected runtime.

Reply:
> I assume that you will want to have libicu in this list as well. Is that correct?

I didn't need to add it to get this compiled on my machine, and `ldd` doesn't show it on my Fedora .NET `so` files. source-build may be using icu in the same way as the portable builds.

Reply:
> Similar to choosing between the .NET bundled runtime packs or the portable Microsoft NuGet packages, the UX for this is the user specify the non-portable rid (instead of the portable rid).

I'm taking a look at the packages involved in a native aot publish using the preview 5 SDK.
I see `Microsoft.DotNet.ILCompiler` which is the tooling, and `runtime.linux-x64.Microsoft.DotNet.ILCompiler` which contains the runtime.

For source-build, the first should be identical to the Microsoft one.
And for the second, we want to produce a non-portable rid package, for example: `runtime.fedora.38-x64.Microsoft.DotNet.ILCompiler`

The extra linker flags describe what is in the second package, so we should include them in that package when it gets built from source.

The first package then needs a mechanism to pick those up.

@jkotas does this make sense?

Reply:
> For source-build, the first should be identical to the Microsoft one.

`runtime.json` file in `Microsoft.DotNet.ILCompiler` package has a list of all RID specific packages that exist. runtime.json` is necessary to make enable restoring of the RID specific package by referencing RID neutral package.

I do not see a good way to include all possible RIDs in the RID neutral package. It means that the RID neutral package produced by source build needs to be different from the Microsoft one; or some of the experiences need to work differently with non-portable source build output.

This problem is not unique to this package. For example, System.IO.Ports package built in dotnet/runtime has the same problem.

Reply:
The split between RID neutral and RID specific packages is only a download size optimization. It is not designed to allow composing RID specific packages of different origin.

Reply:
I think the SDK finds the rid specific package explicitly based on the information from BundledVersions.props which we control for source-build:
```
    <KnownILCompilerPack Include="Microsoft.DotNet.ILCompiler"
                         TargetFramework="net8.0"
                         ILCompilerPackNamePattern="runtime.**RID**.Microsoft.DotNet.ILCompiler"
                         ILCompilerPackVersion="8.0.0-preview.5.23280.8"
                         ILCompilerRuntimeIdentifiers="linux-musl-x64;linux-x64;win-x64;linux-arm;linux-arm64;linux-musl-arm;linux-musl-arm64;osx-arm64;osx-x64;win-arm;win-arm64;win-x86"
                         />
```

Reply:
Yes and `ILCompilerRuntimeIdentifiers` has hardcoded finite lists of RIDs. What happens for RIDs outside this hardcoded list?

Reply:
> Yes and ILCompilerRuntimeIdentifiers has hardcoded finite lists of RIDs. What happens for RIDs outside this hardcoded list?

For some of the lists it will use the runtime graph to find the most appropriate rid from the finite set.

It's clear some changes are needed for the SDK, `Microsoft.DotNet.ILCompiler`, and `runtime.<rid>.Microsoft.DotNet.ILCompiler` to work together when built from source.

I'm going to avoid the linker flags issue for now by making the build use the openssl portable `dlopen` when source-build and nativeaot are combined, and see what else comes up.

Reply:
> I'm going to avoid the linker flags issue for now by making the build use the openssl portable dlopen when source-build and nativeaot are combined, and see what else comes up.

Added https://github.com/dotnet/runtime/pull/88520/commits/30242120a43ad50f68cf6fbe960561d3488b5b0d for this.